File: merger/orpc.go
package merger

import (
	"errors"
	"fmt"
	"sort"

	"github.com/xraph/farp"
)

// ORPCSpec represents a simplified oRPC specification.
type ORPCSpec struct {
	ORPC            string                        `json:"orpc"`
	Info            Info                          `json:"info"`
	Servers         []Server                      `json:"servers,omitempty"`
	Procedures      map[string]ORPCProcedure      `json:"procedures"`
	Schemas         map[string]any                `json:"schemas,omitempty"`
	SecuritySchemes map[string]ORPCSecurityScheme `json:"securitySchemes,omitempty"`
	Security        []map[string][]string         `json:"security,omitempty"`
	Extensions      map[string]any                `json:"-"`
}

// ORPCSecurityScheme represents an oRPC security scheme.
type ORPCSecurityScheme struct {
	Type             string         `json:"type"` // apiKey, http, oauth2, openIdConnect
	Description      string         `json:"description,omitempty"`
	Name             string         `json:"name,omitempty"`             // For apiKey
	In               string         `json:"in,omitempty"`               // For apiKey: header, query, cookie
	Scheme           string         `json:"scheme,omitempty"`           // For http
	BearerFormat     string         `json:"bearerFormat,omitempty"`     // For http bearer
	OpenIdConnectURL string         `json:"openIdConnectUrl,omitempty"` // For openIdConnect
	Flows            map[string]any `json:"flows,omitempty"`            // For oauth2
}

// ORPCProcedure represents an oRPC procedure definition.
type ORPCProcedure struct {
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	Input       *ORPCSchema    `json:"input,omitempty"`
	Output      *ORPCSchema    `json:"output,omitempty"`
	Errors      []ORPCError    `json:"errors,omitempty"`
	Streaming   bool           `json:"streaming,omitempty"`
	Batch       bool           `json:"batch,omitempty"`
	Options     map[string]any `json:"options,omitempty"`
	Extensions  map[string]any `json:"-"`
}

// ORPCSchema represents a schema reference or inline schema.
type ORPCSchema struct {
	Ref    string         `json:"$ref,omitempty"`
	Type   string         `json:"type,omitempty"`
	Schema map[string]any `json:"schema,omitempty"`
}

// ORPCError represents an error definition.
type ORPCError struct {
	Code        int            `json:"code"`
	Message     string         `json:"message"`
	Description string         `json:"description,omitempty"`
	Schema      map[string]any `json:"schema,omitempty"`
}

// ORPCServiceSchema wraps an oRPC schema with its service context.
type ORPCServiceSchema struct {
	Manifest *farp.SchemaManifest
	Schema   any
	Parsed   *ORPCSpec
}

// ORPCMerger handles oRPC schema composition.
type ORPCMerger struct {
	config MergerConfig
}

// NewORPCMerger creates a new oRPC merger.
func NewORPCMerger(config MergerConfig) *ORPCMerger {
	return &ORPCMerger{
		config: config,
	}
}

// ORPCMergeResult contains the merged oRPC spec and metadata.
type ORPCMergeResult struct {
	Spec             *ORPCSpec
	IncludedServices []string
	ExcludedServices []string
	Conflicts        []Conflict
	Warnings         []string
}

// MergeORPC merges multiple oRPC schemas from service manifests.
func (m *ORPCMerger) MergeORPC(schemas []ORPCServiceSchema) (*ORPCMergeResult, error) {
	result := &ORPCMergeResult{
		Spec: &ORPCSpec{
			ORPC: "1.0.0",
			Info: Info{
				Title:       m.config.MergedTitle,
				Description: m.config.MergedDescription,
				Version:     m.config.MergedVersion,
			},
			Servers:    m.config.Servers,
			Procedures: make(map[string]ORPCProcedure),
			Schemas:    make(map[string]any),
			Extensions: make(map[string]any),
		},
		IncludedServices: []string{},
		ExcludedServices: []string{},
		Conflicts:        []Conflict{},
		Warnings:         []string{},
	}

	// Track what we've seen for conflict detection
	seenProcedures := make(map[string]string)      // procedure -> service
	seenSchemas := make(map[string]string)         // schema -> service
	seenSecuritySchemes := make(map[string]string) // security scheme -> service

	// Process each schema
	for _, schema := range schemas {
		serviceName := schema.Manifest.ServiceName

		// Check if this schema should be included
		if !shouldIncludeORPCInMerge(schema) {
			result.ExcludedServices = append(result.ExcludedServices, serviceName)

			continue
		}

		result.IncludedServices = append(result.IncludedServices, serviceName)

		// Parse the schema if not already parsed
		if schema.Parsed == nil {
			parsed, err := ParseORPCSchema(schema.Schema)
			if err != nil {
				result.Warnings = append(result.Warnings,
					fmt.Sprintf("Failed to parse oRPC schema for %s: %v", serviceName, err))

				continue
			}

			schema.Parsed = parsed
		}

		// Get composition config
		compConfig := getORPCCompositionConfig(schema.Manifest)
		strategy := m.getConflictStrategy(compConfig)

		// Determine prefixes
		procedurePrefix := getORPCProcedurePrefix(schema.Manifest, compConfig)
		schemaPrefix := getORPCSchemaPrefix(schema.Manifest, compConfig)

		// Merge procedures
		for procName, procedure := range schema.Parsed.Procedures {
			prefixedName := procName
			if procedurePrefix != "" {
				prefixedName = procedurePrefix + "." + procName
			}

			// Check for procedure conflicts
			if existingService, exists := seenProcedures[prefixedName]; exists {
				conflict := Conflict{
					Type:     "orpc_procedure",
					Item:     procName,
					Services: []string{existingService, serviceName},
					Strategy: strategy,
				}

				switch strategy {
				case farp.ConflictStrategyError:
					return nil, fmt.Errorf("oRPC procedure conflict: %s exists in both %s and %s",
						procName, existingService, serviceName)

				case farp.ConflictStrategySkip:
					conflict.Resolution = "Skipped procedure from " + serviceName
					result.Conflicts = append(result.Conflicts, conflict)

					continue

				case farp.ConflictStrategyOverwrite:
					conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)

				case farp.ConflictStrategyPrefix:
					prefixedName = serviceName + "." + procName
					conflict.Resolution = "Prefixed to " + prefixedName
					result.Conflicts = append(result.Conflicts, conflict)
				}
			}

			result.Spec.Procedures[prefixedName] = procedure
			seenProcedures[prefixedName] = serviceName
		}

		// Merge schemas
		for schemaName, schemaObj := range schema.Parsed.Schemas {
			prefixedName := schemaName
			if schemaPrefix != "" {
				prefixedName = schemaPrefix + "_" + schemaName
			}

			if existingService, exists := seenSchemas[prefixedName]; exists {
				if strategy == farp.ConflictStrategySkip {
					result.Conflicts = append(result.Conflicts, Conflict{
						Type:       ConflictTypeComponent,
						Item:       schemaName,
						Services:   []string{existingService, serviceName},
						Resolution: "Skipped schema from " + serviceName,
						Strategy:   strategy,
					})

					continue
				}
			}

			result.Spec.Schemas[prefixedName] = schemaObj
			seenSchemas[prefixedName] = serviceName
		}

		// Merge security schemes
		for name, secScheme := range schema.Parsed.SecuritySchemes {
			if existingService, exists := seenSecuritySchemes[name]; exists {
				conflict := Conflict{
					Type:     ConflictTypeSecurityScheme,
					Item:     name,
					Services: []string{existingService, serviceName},
					Strategy: strategy,
				}

				switch strategy {
				case farp.ConflictStrategyError:
					return nil, fmt.Errorf("oRPC security scheme conflict: %s exists in both %s and %s",
						name, existingService, serviceName)

				case farp.ConflictStrategySkip:
					conflict.Resolution = "Skipped security scheme from " + serviceName
					result.Conflicts = append(result.Conflicts, conflict)

					continue

				case farp.ConflictStrategyOverwrite:
					conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)

				case farp.ConflictStrategyPrefix:
					prefixedName := serviceName + "_" + name
					conflict.Resolution = "Prefixed to " + prefixedName
					result.Conflicts = append(result.Conflicts, conflict)
					result.Spec.SecuritySchemes[prefixedName] = secScheme
					seenSecuritySchemes[prefixedName] = serviceName

					continue

				case farp.ConflictStrategyMerge:
					conflict.Resolution = fmt.Sprintf("Merged (overwritten) with %s version", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)
				}
			}

			result.Spec.SecuritySchemes[name] = secScheme
			seenSecuritySchemes[name] = serviceName
		}
	}

	return result, nil
}

// ParseORPCSchema parses a raw oRPC schema into structured format.
func ParseORPCSchema(raw any) (*ORPCSpec, error) {
	schemaMap, ok := raw.(map[string]any)
	if !ok {
		return nil, errors.New("schema must be a map")
	}

	spec := &ORPCSpec{
		ORPC:            "1.0.0",
		Procedures:      make(map[string]ORPCProcedure),
		Schemas:         make(map[string]any),
		SecuritySchemes: make(map[string]ORPCSecurityScheme),
		Extensions:      make(map[string]any),
	}

	// Parse oRPC version
	if v, ok := schemaMap["orpc"].(string); ok {
		spec.ORPC = v
	} else {
		spec.ORPC = "1.0.0" // Default
	}

	// Parse info
	if info, ok := schemaMap["info"].(map[string]any); ok {
		spec.Info = parseInfo(info)
	}

	// Parse servers
	if servers, ok := schemaMap["servers"].([]any); ok {
		spec.Servers = parseServers(servers)
	}

	// Parse procedures
	if procedures, ok := schemaMap["procedures"].(map[string]any); ok {
		spec.Procedures = parseORPCProcedures(procedures)
	}

	// Parse schemas
	if schemas, ok := schemaMap["schemas"].(map[string]any); ok {
		spec.Schemas = schemas
	}

	// Parse security schemes
	if securitySchemes, ok := schemaMap["securitySchemes"].(map[string]any); ok {
		spec.SecuritySchemes = parseORPCSecuritySchemes(securitySchemes)
	}

	// Parse global security
	if security, ok := schemaMap["security"].([]any); ok {
		spec.Security = parseSecurityRequirements(security)
	}

	return spec, nil
}

func parseORPCProcedures(procedures map[string]any) map[string]ORPCProcedure {
	result := make(map[string]ORPCProcedure)

	for name, proc := range procedures {
		if procMap, ok := proc.(map[string]any); ok {
			procedure := ORPCProcedure{
				Name:       name,
				Extensions: make(map[string]any),
			}
			if desc, ok := procMap["description"].(string); ok {
				procedure.Description = desc
			}

			if streaming, ok := procMap["streaming"].(bool); ok {
				procedure.Streaming = streaming
			}

			if batch, ok := procMap["batch"].(bool); ok {
				procedure.Batch = batch
			}
			// Parse input/output schemas (simplified)
			result[name] = procedure
		}
	}

	return result
}

func parseORPCSecuritySchemes(schemes map[string]any) map[string]ORPCSecurityScheme {
	result := make(map[string]ORPCSecurityScheme)

	for name, s := range schemes {
		if schemeMap, ok := s.(map[string]any); ok {
			scheme := ORPCSecurityScheme{}
			if t, ok := schemeMap["type"].(string); ok {
				scheme.Type = t
			}

			if desc, ok := schemeMap["description"].(string); ok {
				scheme.Description = desc
			}

			if n, ok := schemeMap["name"].(string); ok {
				scheme.Name = n
			}

			if in, ok := schemeMap["in"].(string); ok {
				scheme.In = in
			}

			if scheme_, ok := schemeMap["scheme"].(string); ok {
				scheme.Scheme = scheme_
			}

			if bf, ok := schemeMap["bearerFormat"].(string); ok {
				scheme.BearerFormat = bf
			}

			if oidc, ok := schemeMap["openIdConnectUrl"].(string); ok {
				scheme.OpenIdConnectURL = oidc
			}

			if flows, ok := schemeMap["flows"].(map[string]any); ok {
				scheme.Flows = flows
			}

			result[name] = scheme
		}
	}

	return result
}

func parseSecurityRequirements(security []any) []map[string][]string {
	result := make([]map[string][]string, 0, len(security))

	for _, req := range security {
		if reqMap, ok := req.(map[string]any); ok {
			requirement := make(map[string][]string)

			for schemeName, scopes := range reqMap {
				if scopeList, ok := scopes.([]any); ok {
					scopeStrings := make([]string, 0, len(scopeList))

					for _, scope := range scopeList {
						if s, ok := scope.(string); ok {
							scopeStrings = append(scopeStrings, s)
						}
					}

					requirement[schemeName] = scopeStrings
				}
			}

			result = append(result, requirement)
		}
	}

	return result
}

// Helper functions for oRPC composition

func shouldIncludeORPCInMerge(schema ORPCServiceSchema) bool {
	for _, schemaDesc := range schema.Manifest.Schemas {
		if schemaDesc.Type == farp.SchemaTypeORPC {
			return true
		}
	}

	return false
}

func getORPCCompositionConfig(manifest *farp.SchemaManifest) *farp.CompositionConfig {
	// oRPC doesn't have composition config yet in spec
	return nil
}

func (m *ORPCMerger) getConflictStrategy(config *farp.CompositionConfig) farp.ConflictStrategy {
	if config != nil && config.ConflictStrategy != "" {
		return config.ConflictStrategy
	}

	return m.config.DefaultConflictStrategy
}

func getORPCProcedurePrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.ComponentPrefix != "" {
		return config.ComponentPrefix
	}

	return manifest.ServiceName
}

func getORPCSchemaPrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.ComponentPrefix != "" {
		return config.ComponentPrefix
	}

	return manifest.ServiceName
}

// SortProcedures sorts procedure names alphabetically.
func SortProcedures(procedures map[string]ORPCProcedure) []string {
	keys := make([]string, 0, len(procedures))
	for k := range procedures {
		keys = append(keys, k)
	}

	sort.Strings(keys)

	return keys
}

File: farp-rust/src/merger/orpc.rs
//! oRPC schema merger for combining multiple service schemas

use super::*;
use crate::errors::Result;
use crate::types::{SchemaManifest, SchemaType};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// oRPC specification
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ORPCSpec {
    pub orpc: String,
    pub info: super::types::Info,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub servers: Vec<super::types::Server>,
    pub procedures: HashMap<String, ORPCProcedure>,
    #[serde(skip_serializing_if = "HashMap::is_empty", default)]
    pub schemas: HashMap<String, serde_json::Value>,
    #[serde(
        skip_serializing_if = "HashMap::is_empty",
        default,
        rename = "securitySchemes"
    )]
    pub security_schemes: HashMap<String, ORPCSecurityScheme>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub security: Vec<HashMap<String, Vec<String>>>,
    #[serde(flatten)]
    pub extensions: HashMap<String, serde_json::Value>,
}

/// oRPC security scheme
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ORPCSecurityScheme {
    #[serde(rename = "type")]
    pub scheme_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "in")]
    pub in_: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scheme: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "bearerFormat")]
    pub bearer_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "openIdConnectUrl")]
    pub openid_connect_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub flows: Option<HashMap<String, serde_json::Value>>,
}

/// oRPC procedure definition
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ORPCProcedure {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub errors: Vec<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub streaming: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<HashMap<String, serde_json::Value>>,
    #[serde(flatten)]
    pub extensions: HashMap<String, serde_json::Value>,
}

/// Service schema with oRPC context
#[derive(Debug, Clone)]
pub struct ORPCServiceSchema {
    pub manifest: SchemaManifest,
    pub schema: serde_json::Value,
    pub parsed: Option<ORPCSpec>,
}

/// oRPC merger
pub struct ORPCMerger {
    config: MergerConfig,
}

/// oRPC merge result
#[derive(Debug, Clone)]
pub struct ORPCMergeResult {
    pub spec: ORPCSpec,
    pub included_services: Vec<String>,
    pub excluded_services: Vec<String>,
    pub conflicts: Vec<Conflict>,
    pub warnings: Vec<String>,
}

impl ORPCMerger {
    pub fn new(config: MergerConfig) -> Self {
        Self { config }
    }

    pub fn merge(&self, schemas: Vec<ORPCServiceSchema>) -> Result<ORPCMergeResult> {
        let mut result = ORPCMergeResult {
            spec: ORPCSpec {
                orpc: "1.0.0".to_string(),
                info: super::types::Info {
                    title: self.config.merged_title.clone(),
                    description: Some(self.config.merged_description.clone()),
                    version: self.config.merged_version.clone(),
                    terms_of_service: None,
                    contact: None,
                    license: None,
                    extensions: HashMap::new(),
                },
                servers: self.config.servers.clone(),
                procedures: HashMap::new(),
                schemas: HashMap::new(),
                security_schemes: HashMap::new(),
                security: Vec::new(),
                extensions: HashMap::new(),
            },
            included_services: Vec::new(),
            excluded_services: Vec::new(),
            conflicts: Vec::new(),
            warnings: Vec::new(),
        };

        let mut seen_procedures: HashMap<String, String> = HashMap::new();
        let mut seen_schemas: HashMap<String, String> = HashMap::new();
        let mut seen_security_schemes: HashMap<String, String> = HashMap::new();

        for mut schema in schemas {
            let service_name = schema.manifest.service_name.clone();

            if !should_include_orpc(&schema) {
                result.excluded_services.push(service_name);
                continue;
            }

            result.included_services.push(service_name.clone());

            if schema.parsed.is_none() {
                match parse_orpc_schema(&schema.schema) {
                    Ok(parsed) => schema.parsed = Some(parsed),
                    Err(e) => {
                        result.warnings.push(format!(
                            "Failed to parse oRPC schema for {service_name}: {e}"
                        ));
                        continue;
                    }
                }
            }

            let parsed = schema.parsed.as_ref().unwrap();
            let strategy = self.config.default_conflict_strategy;

            let procedure_prefix = &schema.manifest.service_name;
            let schema_prefix = &schema.manifest.service_name;

            // Merge procedures
            for (proc_name, procedure) in &parsed.procedures {
                let mut prefixed_name = format!("{procedure_prefix}.{proc_name}");

                if let Some(existing_service) = seen_procedures.get(&prefixed_name) {
                    let conflict = Conflict {
                        conflict_type: ConflictType::Component,
                        item: proc_name.clone(),
                        services: vec![existing_service.clone(), service_name.clone()],
                        resolution: String::new(),
                        strategy,
                    };

                    match strategy {
                        ConflictStrategy::Error => {
                            return Err(crate::errors::Error::Custom(format!(
                                "oRPC procedure conflict: {proc_name} exists in both {existing_service} and {service_name}"
                            )));
                        }
                        ConflictStrategy::Skip => {
                            let mut c = conflict;
                            c.resolution = format!("Skipped procedure from {service_name}");
                            result.conflicts.push(c);
                            continue;
                        }
                        ConflictStrategy::Overwrite => {
                            let mut c = conflict;
                            c.resolution = format!("Overwritten with {service_name} version");
                            result.conflicts.push(c);
                        }
                        ConflictStrategy::Prefix => {
                            prefixed_name = format!("{service_name}.{proc_name}");
                            let mut c = conflict;
                            c.resolution = format!("Prefixed to {prefixed_name}");
                            result.conflicts.push(c);
                        }
                        ConflictStrategy::Merge => {
                            let mut c = conflict;
                            c.resolution = "Merged".to_string();
                            result.conflicts.push(c);
                        }
                    }
                }

                result
                    .spec
                    .procedures
                    .insert(prefixed_name.clone(), procedure.clone());
                seen_procedures.insert(prefixed_name, service_name.clone());
            }

            // Merge schemas
            for (schema_name, schema_obj) in &parsed.schemas {
                let prefixed_name = format!("{schema_prefix}_{schema_name}");
                if let Some(existing_service) = seen_schemas.get(&prefixed_name) {
                    if strategy == ConflictStrategy::Skip {
                        result.conflicts.push(Conflict {
                            conflict_type: ConflictType::Component,
                            item: schema_name.clone(),
                            services: vec![existing_service.clone(), service_name.clone()],
                            resolution: format!("Skipped schema from {service_name}"),
                            strategy,
                        });
                        continue;
                    }
                }

                result
                    .spec
                    .schemas
                    .insert(prefixed_name.clone(), schema_obj.clone());
                seen_schemas.insert(prefixed_name, service_name.clone());
            }

            // Merge security schemes
            for (name, sec_scheme) in &parsed.security_schemes {
                if let Some(existing_service) = seen_security_schemes.get(name) {
                    let conflict = Conflict {
                        conflict_type: ConflictType::SecurityScheme,
                        item: name.clone(),
                        services: vec![existing_service.clone(), service_name.clone()],
                        resolution: String::new(),
                        strategy,
                    };

                    match strategy {
                        ConflictStrategy::Error => {
                            return Err(crate::errors::Error::Custom(format!(
                                "oRPC security scheme conflict: {name} exists in both {existing_service} and {service_name}"
                            )));
                        }
                        ConflictStrategy::Skip => {
                            let mut c = conflict;
                            c.resolution = format!("Skipped security scheme from {service_name}");
                            result.conflicts.push(c);
                            continue;
                        }
                        ConflictStrategy::Overwrite => {
                            let mut c = conflict;
                            c.resolution = format!("Overwritten with {service_name} version");
                            result.conflicts.push(c);
                        }
                        ConflictStrategy::Prefix => {
                            let prefixed_name = format!("{service_name}_{name}");
                            let mut c = conflict;
                            c.resolution = format!("Prefixed to {prefixed_name}");
                            result.conflicts.push(c);
                            result
                                .spec
                                .security_schemes
                                .insert(prefixed_name.clone(), sec_scheme.clone());
                            seen_security_schemes.insert(prefixed_name, service_name.clone());
                            continue;
                        }
                        ConflictStrategy::Merge => {
                            let mut c = conflict;
                            c.resolution =
                                format!("Merged (overwritten) with {service_name} version");
                            result.conflicts.push(c);
                        }
                    }
                }

                result
                    .spec
                    .security_schemes
                    .insert(name.clone(), sec_scheme.clone());
                seen_security_schemes.insert(name.clone(), service_name.clone());
            }
        }

        Ok(result)
    }
}

/// Parse oRPC schema from JSON
pub fn parse_orpc_schema(raw: &serde_json::Value) -> Result<ORPCSpec> {
    serde_json::from_value(raw.clone()).map_err(|e| {
        crate::errors::Error::invalid_schema(format!("Failed to parse oRPC schema: {e}"))
    })
}

fn should_include_orpc(schema: &ORPCServiceSchema) -> bool {
    schema
        .manifest
        .schemas
        .iter()
        .any(|s| s.schema_type == SchemaType::ORPC)
}
